### ISSUE TITLE
Getting panic error when trying to view commit full details

### ISSUE THREAD

**Issue by schpet** [MitMaro/git-interactive-rebase-tool#754]
hi there! just installed this on my mac via homebrew, started a rebase, looked at a commit (c) and then tried to see the full info (d) and got a crash:

```
thread 'main' panicked at 'byte index 21 is not a char boundary; it is inside...
```

version info:

- git version 2.35.1
- interactive-rebase-tool 2.2.0 (2022-05-01)

<img width="1226" alt="Screen Shot 2022-10-12 at 1 49 56 PM" src="https://user-images.githubusercontent.com/921605/195446080-ea04241e-df39-4376-b3c4-35ecf5eee8c7.png">

it seems specific to the one commit i was looking at, others seem to work ok!

unfortunately the repo is not open source, but if there's any help i can do get a stack trace or something i'm happy to. seems like a handy tool so i'm hoping i can use it.

**Comment by MitMaro** [comment]
Hmm, my initial guess is non-UTF data and/or binary files being treated as UTF. 

The release versions do not contain the debug symbols needed to provide a backtrace.

If you are okay trying a development build, you can get a development build for the [latest release](https://github.com/MitMaro/git-interactive-rebase-tool/releases/tag/latest). Then:

```bash
RUST_BACKTRACE=1 GIT_SEQUENCE_EDITOR=/path/to/downloaded-version git rebase -i ...
```

A backtrace would be a big help to narrow down the cause of the crash.

Also, to fix your terminal when this happens:

Type `reset` and then hit ctrl+j. `reset` will fix the invalid state the tool will leave your terminal on crash. And `ctrl+j` is a shortcut for "enter", which usually doesn't work after a crash. Should save you the trouble of having to restart your terminal.

**Comment by MitMaro** [comment]
Ah! Nevermind! I was able to reproduce locally from your screenshot.

![image](https://user-images.githubusercontent.com/177427/195455191-7a45bc6b-aae2-4336-888c-16071a808456.png)


Turns out the `…` in a file causes a crash.

**Comment by schpet** [comment]
oh wow! yeah i love those ellipses… thanks so much for checking this out, and debugging that saving me from figuring out how to build this thing.

**Comment by MitMaro** [comment]
So! Turns out that you discovered a tricky bug that has slipped through the cracks, and has existed for a while now. It is also a complete miss and incorrect usage of an external library on my part. Thanks for the report!

Essentially, any character that takes more than 1 byte, and is a leading or trailing character, excluding whitespace characters, will cause a crash.

I have added some tests as well as a fix, and hope to have a PR up soon. Since I've made a number of changes to the project since the 2.2.0 release, and am not ready to release some of the more recent changes, I will also backport the fix to 2.2.0 and release 2.2.1,

---

Technical details (if you are curious)

To show tailing and leading whitespace characters, I do a search of the start and end of each line and retrieve the index of the first non-whitespace character and the index of the last non-whitespace character.

To find the end index, I was doing:

```rust
for (index, c) in graphemes.rev().enumerate() {
	if c != " " && c != "\t" && c != "\n" {
		end_partition_index = length - index;
		break;
	}

```

The problem is that `index` isn't the byte position in the string value, but the character index. So in the case of `…` (`\xE2\x80\xA6`), I was returning an "end" index value of 1 instead of 3. (Note: I add 1 to the index because of the slice syntax in rust being exclusive on the ending index)

This meant that when trying to get a substring value of `…`, I was only grabbing the first byte (`\xE2`). Since `\xE2` isn't a valid UTF-8 it caused the panic. 

I, for some unknown reason, had made the assumption that the [UnicodeSegmentation library](https://docs.rs/unicode-segmentation/latest/unicode_segmentation/) returned the "byte index" when enumerating, instead of the index in the characters "list". 😞

**Comment by schpet** [comment]
oh wow, i really enjoyed reading through this explanation! thanks so much for providing it. i did enjoy the way the program dramatically crashed, with my terminal beeping and the whole thing unresponsive, but getting this fix is something i look forward to as well! will keep an eye on the releases!

**Comment by MitMaro** [comment]
Just released [version 2.2.1](https://github.com/MitMaro/git-interactive-rebase-tool/releases/tag/2.2.1) with the fix.

If you are using homebrew, it usually gets updated fairly quickly through an automated process on their end.

If you are curious on the change, PR #755 fixes the problem in the development build, and if identical.

Thanks again for the report!
